README
This is Perl module Archive::Zip::Uncompressed.

INSTALLATION

Archive::Zip::Uncompressed installation is straightforward. If your CPAN shell is set up,
you should just be able to do

    % cpan Archive::Zip::Uncompressed

Download it, unpack it, then build it as per the usual:

    % perl Makefile.PL
    % make && make test

Then install it:

    % make install

DOCUMENTATION

Archive::Zip::Uncompressed documentation is available as in POD. So you can do:

    % perldoc Archive::Zip::Uncompressed

to read the documentation online with your favorite pager.

Tokuhiro Matsuno


Changes
Revision history for Perl extension Archive::Zip::Uncompressed

0.01    Thu Feb 25 12:42:23 2010
        - original version


lib/Archive/Zip/Uncompressed.pm
package Archive::Zip::Uncompressed;
use strict;
use warnings;
use 5.00800;
use Archive::Zip::Uncompressed::Unpacker;
use Archive::Zip::Uncompressed::Packer;

our $VERSION = '0.01';
use Carp ();

our $BUFFER_SIZE = 1024*1024;

1;
__END__

=encoding utf8

=head1 NAME

Archive::Zip::Uncompressed -

=head1 SYNOPSIS

    use strict;
    use warnings;
    use Archive::Zip::Uncompressed;

    my $ofname = shift @ARGV or die "Usage: $0 foo.zip files";
    my $zip = Archive::Zip::Uncompressed::Packer->new($ofname);
    for my $file (@ARGV) {
        if (-d $file) {
            $file =~ s{([^/])$}{$1/};
            $zip->add_directory($file);
        } else {
            $zip->add_file($file);
        }
    }


=head1 DESCRIPTION

Archive::Zip::Uncompressed is uncompressed zip generator for Perl.
This module is useful for packing JPEG, MP3, and other compressed data.

=head1 FAQ

=over 4

=item Do you know Archive::Zip?

Yes.

L<Archive::Zip> is awesome library for handling zip archive.
But, the module has dependencies for XS.

=back

=head1 AUTHOR

Tokuhiro Matsuno E<lt>tokuhirom AAJKLFJEF GMAIL COME<gt>

=head1 SEE ALSO

L<http://www.pkware.com/documents/casestudies/APPNOTE.TXT>, L<Archive::Zip>

=head1 LICENSE

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut


lib/Archive/Zip/Uncompressed/Packer.pm
package Archive::Zip::Uncompressed::Packer;
use strict;
use warnings;
use File::stat;
use Fcntl ':seek';
use constant {
    'HeaderMaker' => 'Archive::Zip::Uncompressed::HeaderMaker',
};

sub new {
    my ($class, $file) = @_;
    my $fh;
    if (ref $file eq 'GLOB' || ref $file eq 'IO') {
        $fh = $file;
    } else {
        open $fh, '>', $file or die "Cannot open archive file: $!";
    }

    bless {
        fh              => $fh,
        central_headers => [],
    }, $class;
}

sub add_file {
    my ($self, $fname) = @_;
    my $header_pos = tell($self->{fh});
    Carp::croak("tell: $fname, $!") if $header_pos == -1;
    my $stat = stat($fname) or die "Cannot call stat: '$fname'";
    open my $ifh, '<', $fname or Carp::croak("Cannot open file: $fname");
    my $crc32 = Archive::Zip::Uncompressed::Packer::CRC32->calc_crc32_from_file($ifh);
    seek($ifh, 0, SEEK_SET) or Carp::croak("Cannot seek file: $fname");
    my $header_body = Archive::Zip::Uncompressed::HeaderMaker->make_file_header_body(
        $stat->mtime,
        $crc32,
        $stat->size,
        length($fname),
    );
    print {$self->{fh}} Archive::Zip::Uncompressed::HeaderMaker->make_file_header($header_body);
    print {$self->{fh}} $fname;
    my $read;
    while ($read = read($ifh, my $buf, $Archive::Zip::Uncompressed::BUFFER_SIZE)) {
        print {$self->{fh}} $buf;
    }
    Carp::croak("cannot read from $fname: $!") unless defined $read;
    push @{$self->{central_headers}}, Archive::Zip::Uncompressed::HeaderMaker->make_central_directory_file_header(
        $header_body,
        Archive::Zip::Uncompressed::HeaderMaker::EXTERNAL_ATTR_DIRECTORY(),
        $header_pos,
        $fname,
    );
    return;
}

sub add_directory {
    my ($self, $fname) = @_;
    my $header_pos = tell($self->{fh});
    Carp::croak("tell: $fname, $!") if $header_pos == -1;
    $fname .= '/' unless $fname =~ m{/$}; # XXX this is really required!

    my $stat = stat($fname) or die "Cannot call stat: '$fname'";
    my $crc32 = 0;
    my $header_body = Archive::Zip::Uncompressed::HeaderMaker->make_file_header_body(
        $stat->mtime,
        $crc32,
        0, # size
        length($fname),
    );
    print {$self->{fh}} Archive::Zip::Uncompressed::HeaderMaker->make_file_header($header_body);
    print {$self->{fh}} $fname;
    push @{$self->{central_headers}}, Archive::Zip::Uncompressed::HeaderMaker->make_central_directory_file_header(
        $header_body,
        Archive::Zip::Uncompressed::HeaderMaker::EXTERNAL_ATTR_DIRECTORY(),
        $header_pos,
        $fname,
    );
    return;
}

sub close {
    my $self = shift;
    Carp::croak("do not call close() twice") if $self->{closed}++;

    my $startpos = tell($self->{fh});
    Carp::croak("tell: $!") if $startpos == -1;

    for my $cheader (@{$self->{central_headers}}) {
        print {$self->{fh}} $cheader;
    }

    my $endpos = tell($self->{fh});
    Carp::croak("tell: $!") if $endpos == -1;

    my $end = Archive::Zip::Uncompressed::HeaderMaker->make_end_central_directory_header(
        scalar(@{$self->{central_headers}}),
        $endpos - $startpos,
        $startpos,
    );
    print {$self->{fh}} $end;

    close($self->{fh});
}

sub DESTROY {
    my $self = shift;
    $self->close() unless $self->{closed};
}

package # hide from pause
    Archive::Zip::Uncompressed::HeaderMaker;
use strict;
use warnings;

use constant {
    EXTERNAL_ATTR_DIRECTORY => ((oct('040755') << 16) | 0x10),
    EXTERNAL_ATTR_FILE      => (oct('0100644') << 16),
};

sub pack_16 { pack('v', $_[0]) }
sub pack_32 { pack('V', $_[0]) }

sub make_file_header {
    my ($self, $header_body) = @_;
    join(
        '',
        "PK\x03\x04",          # int signature
        $header_body,
    );
}

# @args $time unix time
sub make_file_header_body {
    my ($class, $time, $crc32, $size, $fnamelen) = @_;
    Carp::croak("too large file") if $size > 0xFFFFFFFF; # really?
    Carp::croak("too long file name") if $fnamelen > 0xFFFF;

    my ($sec,$min,$hour,$mday,$mon,$year,$wday,$yday,$isdst) = localtime($time);
    my $dosdate = (($year+1900-1980)<<9) | ($mon+1)<<5 | $mday;
    my $dostime = ($hour<<11) | ($min<<5) | ($sec>>1);
    my $extralen = 0;

    return join(
        '',
        pack_16(0x14),         # short needver
        pack_16(0),            # short option
        pack_16(0),            # short comptype
        pack_16($dostime),     # short filetime
        pack_16($dosdate),     # short filedate
        pack_32($crc32),       # int crc32
        pack_32($size),        # int compsize
        pack_32($size),        # int uncompsize
        pack_16($fnamelen),    # short fnamelen
        pack_16(0),            # short extralen
    );
}

sub make_central_directory_file_header {
    my ($class, $file_header_body, $external_attr, $header_pos, $fname) = @_;
    return join(
        '',
        "\x50\x4B\x01\x02", # unsigned int signature;
        pack_16(0x0314),    # unsigned short madever;
        $file_header_body,
        pack_16(0), # unsigned short commentlen;
        pack_16(0), # unsigned short disknum;
        pack_16(0), # unsigned short inattr;
        pack_32($external_attr), # unsigned int outattr;
        pack_32($header_pos), # unsigned int headerpos;
        $fname,
    );
}

# $member_num is the number of members
# $dirsize is the sum of length of central directory headers
sub make_end_central_directory_header {
    my ($class, $member_num, $dirsize, $startpos) = @_;
    join(
        '' => (
            "\x50\x4B\x05\x06",    # unsigned int signature;
            pack_16(0),            # unsigned short disknum;
            pack_16(0),            # unsigned short startdisknum;
            pack_16($member_num),  # unsigned short diskdirentry;
            pack_16($member_num),  # unsigned short direntry;
            pack_32($dirsize),     # unsigned int dirsize;
            pack_32($startpos),    # unsigned int startpos;
            pack_16(0),            # unsigned short commentlen;
        )
    );
}

package # hide from pause
    Archive::Zip::Uncompressed::Packer::CRC32;

our @CRCTABLE = (
    0x0,        0x77073096, 0xEE0E612C, 0x990951BA, 0x76DC419,  0x706AF48F,
    0xE963A535, 0x9E6495A3, 0xEDB8832,  0x79DCB8A4, 0xE0D5E91E, 0x97D2D988,
    0x9B64C2B,  0x7EB17CBD, 0xE7B82D07, 0x90BF1D91, 0x1DB71064, 0x6AB020F2,
    0xF3B97148, 0x84BE41DE, 0x1ADAD47D, 0x6DDDE4EB, 0xF4D4B551, 0x83D385C7,
    0x136C9856, 0x646BA8C0, 0xFD62F97A, 0x8A65C9EC, 0x14015C4F, 0x63066CD9,
    0xFA0F3D63, 0x8D080DF5, 0x3B6E20C8, 0x4C69105E, 0xD56041E4, 0xA2677172,
    0x3C03E4D1, 0x4B04D447, 0xD20D85FD, 0xA50AB56B, 0x35B5A8FA, 0x42B2986C,
    0xDBBBC9D6, 0xACBCF940, 0x32D86CE3, 0x45DF5C75, 0xDCD60DCF, 0xABD13D59,
    0x26D930AC, 0x51DE003A, 0xC8D75180, 0xBFD06116, 0x21B4F4B5, 0x56B3C423,
    0xCFBA9599, 0xB8BDA50F, 0x2802B89E, 0x5F058808, 0xC60CD9B2, 0xB10BE924,
    0x2F6F7C87, 0x58684C11, 0xC1611DAB, 0xB6662D3D, 0x76DC4190, 0x1DB7106,
    0x98D220BC, 0xEFD5102A, 0x71B18589, 0x6B6B51F,  0x9FBFE4A5, 0xE8B8D433,
    0x7807C9A2, 0xF00F934,  0x9609A88E, 0xE10E9818, 0x7F6A0DBB, 0x86D3D2D,
    0x91646C97, 0xE6635C01, 0x6B6B51F4, 0x1C6C6162, 0x856530D8, 0xF262004E,
    0x6C0695ED, 0x1B01A57B, 0x8208F4C1, 0xF50FC457, 0x65B0D9C6, 0x12B7E950,
    0x8BBEB8EA, 0xFCB9887C, 0x62DD1DDF, 0x15DA2D49, 0x8CD37CF3, 0xFBD44C65,
    0x4DB26158, 0x3AB551CE, 0xA3BC0074, 0xD4BB30E2, 0x4ADFA541, 0x3DD895D7,
    0xA4D1C46D, 0xD3D6F4FB, 0x4369E96A, 0x346ED9FC, 0xAD678846, 0xDA60B8D0,
    0x44042D73, 0x33031DE5, 0xAA0A4C5F, 0xDD0D7CC9, 0x5005713C, 0x270241AA,
    0xBE0B1010, 0xC90C2086, 0x5768B525, 0x206F85B3, 0xB966D409, 0xCE61E49F,
    0x5EDEF90E, 0x29D9C998, 0xB0D09822, 0xC7D7A8B4, 0x59B33D17, 0x2EB40D81,
    0xB7BD5C3B, 0xC0BA6CAD, 0xEDB88320, 0x9ABFB3B6, 0x3B6E20C,  0x74B1D29A,
    0xEAD54739, 0x9DD277AF, 0x4DB2615,  0x73DC1683, 0xE3630B12, 0x94643B84,
    0xD6D6A3E,  0x7A6A5AA8, 0xE40ECF0B, 0x9309FF9D, 0xA00AE27,  0x7D079EB1,
    0xF00F9344, 0x8708A3D2, 0x1E01F268, 0x6906C2FE, 0xF762575D, 0x806567CB,
    0x196C3671, 0x6E6B06E7, 0xFED41B76, 0x89D32BE0, 0x10DA7A5A, 0x67DD4ACC,
    0xF9B9DF6F, 0x8EBEEFF9, 0x17B7BE43, 0x60B08ED5, 0xD6D6A3E8, 0xA1D1937E,
    0x38D8C2C4, 0x4FDFF252, 0xD1BB67F1, 0xA6BC5767, 0x3FB506DD, 0x48B2364B,
    0xD80D2BDA, 0xAF0A1B4C, 0x36034AF6, 0x41047A60, 0xDF60EFC3, 0xA867DF55,
    0x316E8EEF, 0x4669BE79, 0xCB61B38C, 0xBC66831A, 0x256FD2A0, 0x5268E236,
    0xCC0C7795, 0xBB0B4703, 0x220216B9, 0x5505262F, 0xC5BA3BBE, 0xB2BD0B28,
    0x2BB45A92, 0x5CB36A04, 0xC2D7FFA7, 0xB5D0CF31, 0x2CD99E8B, 0x5BDEAE1D,
    0x9B64C2B0, 0xEC63F226, 0x756AA39C, 0x26D930A,  0x9C0906A9, 0xEB0E363F,
    0x72076785, 0x5005713,  0x95BF4A82, 0xE2B87A14, 0x7BB12BAE, 0xCB61B38,
    0x92D28E9B, 0xE5D5BE0D, 0x7CDCEFB7, 0xBDBDF21,  0x86D3D2D4, 0xF1D4E242,
    0x68DDB3F8, 0x1FDA836E, 0x81BE16CD, 0xF6B9265B, 0x6FB077E1, 0x18B74777,
    0x88085AE6, 0xFF0F6A70, 0x66063BCA, 0x11010B5C, 0x8F659EFF, 0xF862AE69,
    0x616BFFD3, 0x166CCF45, 0xA00AE278, 0xD70DD2EE, 0x4E048354, 0x3903B3C2,
    0xA7672661, 0xD06016F7, 0x4969474D, 0x3E6E77DB, 0xAED16A4A, 0xD9D65ADC,
    0x40DF0B66, 0x37D83BF0, 0xA9BCAE53, 0xDEBB9EC5, 0x47B2CF7F, 0x30B5FFE9,
    0xBDBDF21C, 0xCABAC28A, 0x53B39330, 0x24B4A3A6, 0xBAD03605, 0xCDD70693,
    0x54DE5729, 0x23D967BF, 0xB3667A2E, 0xC4614AB8, 0x5D681B02, 0x2A6F2B94,
    0xB40BBE37, 0xC30C8EA1, 0x5A05DF1B, 0x2D02EF8D
);

sub calc_crc32_from_string {
    my ($class, $src, $result) = @_;
    $result = 0xFFFFFFFF unless defined $result;
    for my $c (split //, $src) {
        $result = ($result >> 8) ^ $CRCTABLE[ord($c) ^ ($result & 0xFF)];
    }
    return (~$result) & 0xFFFFFFFF;
}

sub calc_crc32_from_file {
    my ($class, $fh) = @_;
    my $result = 0xFFFFFFFF;
    my $read;
    while ($read = read($fh, my $buf, $Archive::Zip::Uncompressed::BUFFER_SIZE)) {
        $result = $class->calc_crc32_from_string($buf, $result);
    }
    Carp::croak("cannot read from file: $!") unless defined $read;
    return $result;
}

1;
__END__

=head1 NAME

Archive::Zip::Uncompressed::Packer - uncompressed zip archive packer

=head1 DESCRIPTION

This class is part of Archive::Zip::Uncompressed distribution.

=head1 AUTHOR

Tokuhiro Matsuno



lib/Archive/Zip/Uncompressed/Unpacker.pm
package Archive::Zip::Uncompressed::Unpacker;
use strict;
use warnings;
use File::stat;
use Fcntl ':seek';
use constant {
    HEADER_SIZE         => 30,
    CENTRAL_HEADER_SIZE => 46,
    END_HEADER_SIZE     => 22,
};

sub unpack_16 { unpack 'v', shift }
sub unpack_32 { unpack 'V', shift }

sub new {
    my ($class, $file) = @_;
    my $fh;
    if (ref $file eq 'GLOB' || ref $file eq 'IO') {
        $fh = $file;
    } else {
        open $fh, '<', $file or die "Cannot open archive file: $!";
    }

    # read END header
    my $size = -s $file;
    seek($fh, -(END_HEADER_SIZE), SEEK_END) or Carp::croak("seek to end header: $!");
    read($fh, my $end_cent_buf, END_HEADER_SIZE) == END_HEADER_SIZE or Carp::croak("cannot read end header");
    substr($end_cent_buf, 0, 4) eq "\x50\x4B\x05\x06" or Carp::croak("invalid end-of-central-directory-header");
    my $member_num = unpack_16(substr($end_cent_buf,  8, 2));
    my $dirsize    = unpack_16(substr($end_cent_buf, 12, 4));
    my $startpos   = unpack_16(substr($end_cent_buf, 16, 4));
        #  0 "\x50\x4B\x05\x06",    # unsigned int signature;
        #  4 pack_16(0),            # unsigned short disknum;
        #  6 pack_16(0),            # unsigned short startdisknum;
        #  8 pack_16($member_num),  # unsigned short diskdirentry;
        # 10 pack_16($member_num),  # unsigned short direntry;
        # 12 pack_32($dirsize),     # unsigned int dirsize;
        # 16 pack_32($startpos),    # unsigned int startpos;
        # 20 pack_16(0),            # unsigned short commentlen;

    my $self = bless {
        fh              => $fh,
        size            => $size,
        member_num      => $member_num,
        dirsize         => $dirsize,
        startpos        => $startpos,
        idx             => 0,
        next_central    => $startpos,
    }, $class;
    return $self;
}

sub next {
    my ($self) = @_;
    my $idx = $self->{idx}++;
    return if $idx >= $self->{member_num};

    # seek to head of central header
    seek($self->{fh}, $self->{next_central}, SEEK_SET) or Carp::croak("seek: $!");
    read($self->{fh}, my $buf, CENTRAL_HEADER_SIZE) == CENTRAL_HEADER_SIZE or Carp::croak("cannot seek");
    if ( substr( $buf, 0, 4 ) ne "\x50\x4B\x01\x02" ) {
        Carp::confess(
            sprintf( "invalid header signature: %08X ne %08X",
                unpack_32( "\x50\x4B\x01\x02" ),
                unpack_32( substr( $buf, 0, 4 ) ) )
        );
    }
    my $dosdt = substr( $buf, 12, 4 );
    my $size       = unpack_32( substr( $buf, 20, 4 ) );
    my $fname_len  = unpack_16( substr( $buf, 28, 2 ) );
    my $external_file_attributes   = unpack_32( substr( $buf, 38, 4 ) );
    my $header_pos = unpack_32( substr( $buf, 42, 4 ) );
    read($self->{fh}, my $fname, $fname_len)==$fname_len or Carp::croak("cannot read file name from file");

      # 0  "\x50\x4B\x01\x02", # unsigned int signature;
      # 4  pack_16(0x0314),    # unsigned short madever;
      # 6  pack_16(0x14),         # short needver
      # 8  pack_16(0),            # short option
      # 10 pack_16(0),            # short comptype
      # 12 pack_16($dostime),     # short filetime
      # 14 pack_16($dosdate),     # short filedate
      # 16 pack_32($crc32),       # int crc32
      # 20 pack_32($size),        # int compsize
      # 24 pack_32($size),        # int uncompsize
      # 28 pack_16($fnamelen),    # short fnamelen
      # 30 pack_16(0),            # short extralen
      # 32 pack_16(0), # unsigned short commentlen;
      # 34 pack_16(0), # unsigned short disknum;
      # 36 pack_16(0), # unsigned short inattr;
      # 38 pack_32($external_attr), # unsigned int outattr;
      # 42 pack_32($header_pos), # unsigned int headerpos;

    $self->{next_central} = $self->{next_central} + CENTRAL_HEADER_SIZE + $fname_len;

    return Archive::Zip::Uncompressed::Unpacker::Member->new(
        fh         => $self->{fh},
        external_file_attributes   => $external_file_attributes,
        size       => $size,
        header_pos => $header_pos,
        filename   => $fname,
        fname_len  => $fname_len,
        'pos'      => $header_pos + HEADER_SIZE + $fname_len,
    );
}

package #
    Archive::Zip::Uncompressed::Unpacker::Member;
use Fcntl ':seek';

sub new {
    my $class = shift;
    bless {@_}, $class;
}

# $self->read(my $buf, $bufsiz)
sub read {
    my $self = $_[0];
    seek($self->{fh}, $self->{pos}, SEEK_SET) or Carp::croak("cannot seek: $!");
    my $buflen = $_[2];
    if ( $self->{header_pos} +
        Archive::Zip::Uncompressed::Unpacker::HEADER_SIZE() +
        $self->{fname_len} +
        $self->{size} < $self->{pos} +
        $buflen )
    {
        $buflen =
          ( $self->{header_pos} +
              Archive::Zip::Uncompressed::Unpacker::HEADER_SIZE() +
              $self->{fname_len} +
              $self->{size} ) - $self->{pos};
    }
    read($self->{fh}, $_[1], $buflen);
}

BEGIN {
    no strict 'refs';
    for (qw/external_file_attributes size filename/) {
        my $key = $_;
        *{__PACKAGE__ . '::' . $_ } = sub { $_[0]->{$key} }
    }
}

1;
__END__

=head1 NAME

Archive::Zip::Uncompressed::Unpacker - uncompressed zip archive unpacker

=head1 DESCRIPTION

This class is part of Archive::Zip::Uncompressed distribution.

=head1 AUTHOR

Tokuhiro Matsuno

